FSW/controllers/controller.py

```diff
@@ -0,0 +1,168 @@
+            z_body = -est_world_states[13:16] / np.linalg.norm(est_world_states[13:16])
+            x_body = np.cross(z_body, np.array([0,0,1]))
+            y_body = np.cross(z_body, x_body)
+            Rot_mat = np.vstack((x_body, y_body, z_body)).T
+            q_ref = rotmat2quat(Rot_mat)
+        elif pointing_target == "Nadir":
+            att_states = est_world_states[6:13]
+            # Default values if not sun pointing
+            ref_ctrl_states = np.zeros((19,))
+            ref_torque = np.zeros((3,))
+            
+        if  pointing_mode == "detumble":
+            # Default values if not sun pointing
+            ref_ctrl_states = np.zeros((19,))
+            ref_torque = np.zeros((3,))
+        elif pointing_mode == "spin-stabilized":
+            # Default values if not sun pointing
+            ref_ctrl_states = np.zeros((19,))
+            # reference angular velocity in pointing axis 
+            # ref_ctrl_states[:4]  = q_ref
+            # ref_ctrl_states[4:7] = tgt_ang_vel
+            ref_torque = np.zeros((3,))
+        elif pointing_mode == "3D-stabilized":
+            q_ref = rotmat2quat(Rot_mat)
+            ref_ctrl_states[:4] = q_ref
+            ref_torque = np.zeros((3,))
+        else:
+            raise ValueError(f"Unrecognized pointing mode: {pointing_mode}")
+            
+        return ref_torque, ref_ctrl_states
+    
+    def get_torque(self, date, ref_torque, ref_ctrl_states, est_ctrl_states, Idx):
+        # from the reference and estimated quaternions, get the error quaternion for control
+        if self.controller_algo == "Bcross":
+            mtb_torque = self.Bcrossctr.get_dipole_moment_command(est_ctrl_states, Idx)
+            return mtb_torque, []
+        elif self.controller_algo == "Lyapunov":
+            mtb_torque = self.lyapsunpointctr.get_dipole_moment_command(est_ctrl_states, Idx) 
+            return mtb_torque, []
+        elif self.controller_algo == "BaseSP":
+            mtb_torque = self.basesunpointctr.get_dipole_moment_command(est_ctrl_states, Idx) 
+            return mtb_torque, []
+        elif self.controller_algo == "BaseNP":
+            mtb_torque, rw_torque = self.nadirpointctr.get_dipole_moment_command_and_rw_torque(est_ctrl_states, Idx) 
+            return mtb_torque, rw_torque
+        else: 
+            raise ValueError(f"Unrecognized controller algorithm: {self.controller_algo}")
+        
+    
+    def allocate_torque(self, state, mtb_torque_cmd, rw_torque_cmd, Idx):
+        # Placeholder for actuator management
+        # if torque = B @ actuator_cmd
+        # actuator_cmd = Binv @ torque_cmd
+        
+     
+        self.allocation_mat = np.zeros((Idx["NU"],3))
+        self.allocation_mat[Idx["U"]["MTB_TORQUE"],:] = np.linalg.pinv(self.G_mtb_b.T)
+                
+        # np.linalg.pinv(B_mat)
+        actuator_cmd = np.zeros((Idx["NU"],1))
+        actuator_cmd[Idx["U"]["MTB_TORQUE"]] = np.linalg.pinv(self.G_mtb_b.T) @ mtb_torque_cmd
```

Comment:
I don't know how badly constrained we are with memory, but we should probably just use lists which contain moments in a pre-defined order

Reply:
we can precompute this in the class and load it every time. This isnt the actual flight code, so for these sims in particular its not an issue. For SIL it will be and we will have to do what you describe